piff/polynomial_interp.py

```diff
@@ -0,0 +1,361 @@
+                           
+        :returns:          A guess for the parameters. In this case a 2D matrix
+                           which is zero everywhere except for (0,0).  This should
+                           correspond to a flat function of the parameters with 
+                           value given by the mean.
+
+
+        """        
+        #We need a starting point for the fitter.
+        #Use a constant value over the whole field as 
+        #a reasonable guess.
+        n = self.orders[parameter_index]+1
+        C = numpy.zeros((n,n))
+        C[0,0] = parameter.mean()
+        return C
+
+
+
+    def solve(self, pos, vectors, logger=None):
+        """Solve for the interpolation coefficients given some data,
+        using the scipy.optimize.curve_fit routine, which uses Levenberg-Marquardt
+        to find the least-squares solution.
+
+        This currently assumes that our positions pos are just u and v.
+
+        :param pos:         A list of positions to use for the interpolation.
+        :param vectors:     A list of parameter vectors (numpy arrays) for each star.
+        :param logger:      A logger object for logging debug info. [default: None]
+        """
+        import scipy.optimize
+
+        #We will want to index things later, so useful
+        #to convert these to numpy arrays and transpose
+        #them to the order we need.
+        parameters = numpy.array(vectors).T
+        positions = numpy.array(pos).T
+       
+        #We should have the same number of parameters as number of polynomial 
+        #orders with which we were created here.
+        nparam = len(parameters)
+        assert nparam==self.nparam, """Must create Polynomial interpolator 
+        with the same order as the input vectors ({}!={})""".format(nparam,self.nparam)
+
+        coeffs = []
+
+        #This model function adapts our interpolationModel method
+        #into the form that the scipy curve_fit function is expecting.
+        #It just needs to unpack a linear exploded list of coefficients
+        #into the matrix form that interpolationModel wants.
+
+
+        #Loop through the parameters
+        for i, parameter in enumerate(parameters):
+
+            def model(uv,*coeffs):
+                C = self.unpack_coefficients(i, coeffs)
+                return self.interpolationModel(uv,C)
+
+            #To replace the polynomial function in this code
+            #with another model it should only be necessary to 
+            #override the methods, initial_guess, fit_function,
+            #and perhaps the pack and unpack methods
+            p0 = self.pack_coefficients(i, self.initialGuess(positions, parameter, i))
+            #print(p0)
+            # Black boxes curve fitter from scipy!
+            # We may want to look into the tolerance and other parameters
+            # of this function.
+            p,covmat=scipy.optimize.curve_fit(model, positions, parameter, p0)
```

Comment:
I think in general it ends up being faster / more stable. This was the same thing matt did in osprey.

We should also be able to include the weight map in the interpolation fit, which should just be as simple as:

    # weights are the same shape as parameters
    sigma = np.zeros_like(weights)
    msk = weights > 0
    sigma = np.sqrt(1.0/weights[msk])
    parameter = parameter[msk]
    positions=positions[msk]
    p,covmat=scipy.optimize.curve_fit(model, positions, parameter, p0, sigma=sigma, absolute_sigma=True)